jwala-services/src/main/java/com/cerner/jwala/service/group/impl/GroupServiceImpl.java

```diff
@@ -411,7 +416,13 @@ protected CommandOutput executeDeployGroupAppTemplate(final String groupName, fi
                 .setGroupName(groupName)
                 .setWebAppName(application.getName())
                 .build();
-        return resourceService.generateAndDeployFile(resourceIdentifier, application.getName(), fileName, hostName);
+        String lockKey = groupName + application.getName() + fileName + JwalaUtils.getHostAddress(hostName);
```

Comment:
I think we didn't lock it before because we always assumed that the resource was only being deployed once per host. I was looking at the REST code, and the way it works is that we add the hostname for each JVM in the group to a list, and if the hostname is already in the list then we skip that deploy. So, as long as the hosts are in the same format only one would get deployed per host.

I'm thinking now, that with this change, the resource will be deployed multiple times for the same host if the hostname is specified for different JVMs using the short name, FQDN, and IP address.

Would it make more sense to populate the list of JVM hosts with the derived IP address and test against that instead?

I'm talking about this code in GroupServiceRestImpl.performGroupAppDeployToHosts (lines 769-779):
`            List<String> deployedHosts = new ArrayList<>(jvms.size());
            for (final Jvm jvm : jvms) {
                final String hostName = jvm.getHostName();
                if (!deployedHosts.contains(hostName)) {
                    historyFacadeService.write(hostName, group, "Deploying application resource " + fileName, EventType.USER_ACTION_INFO, aUser.getUser().getId());
                    deployedHosts.add(hostName);
                    Future<Response> response = createFutureResponseForAppDeploy(groupName, fileName, appName, jvm, null);
                    if (response != null)
                        futureMap.put(hostName, response);
                }
            }`